### PR TITLE
#27 implement database manager and connect

### PR DESCRIPTION
**Pull request by kortov**
connect to #27

### CODE REVIEW

**Review by ivshapovalov — APPROVED** [review]
Well done. . Test method names is good.

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        jdbcDriverClass = "org.postgresql.Driver";
+        jdbcUrl = "jdbc:postgresql://localhost:5432/";
+    }
+
+    public DatabaseManager(String jdbcDriverClass, String jdbcUrl) {
+        this.jdbcDriverClass = jdbcDriverClass;
+        this.jdbcUrl = jdbcUrl;
+    }
+
+    boolean connect(String database, String user, String password) {
```
Comment (AVM82) [inline src/main/java/juja/sqlcmd/DatabaseManager.java]: Please make this method public.

**Review by AVM82 — CHANGES_REQUESTED** [review]

**Review by AVM82 — APPROVED** [review]

File: src/main/java/juja/sqlcmd/DatabaseManager.java (hunk 1 above)
Comment (kortov) [reply src/main/java/juja/sqlcmd/DatabaseManager.java]: Done

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 2]
```diff
@@ -5,16 +5,33 @@
...
+    private final String jdbcDriverClass;
+    private final String jdbcUrl;
     private Connection connection;
 
+    public DatabaseManager() {
+        jdbcDriverClass = "org.postgresql.Driver";
+        jdbcUrl = "jdbc:postgresql://localhost:5432/";
+    }
+
+    public DatabaseManager(String jdbcDriverClass, String jdbcUrl) {
```
Comment (Shtramak) [inline src/main/java/juja/sqlcmd/DatabaseManager.java]: Who will use this constructor to create instance of DatabaseManager()? Should client know the details about jdbcDriverClass or jdbcUrl? In this case I think it's better to have jdbcDriverClass and jdbcUrl declared as a constants.

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 3]
```diff
@@ -5,16 +5,33 @@
 import java.sql.SQLException;
 
 public class DatabaseManager {
+    private final String jdbcDriverClass;
+    private final String jdbcUrl;
```
Comment (Shtramak) [inline src/main/java/juja/sqlcmd/DatabaseManager.java]: Pay your attention on whitespaces (http://bit.ly/2j2qPE1)

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 4]
```diff
@@ -5,16 +5,33 @@
...
+        this.jdbcDriverClass = jdbcDriverClass;
+        this.jdbcUrl = jdbcUrl;
+    }
+
     boolean connect(String database, String user, String password) {
         try {
-            Class.forName("org.postgresql.Driver");
+            Class.forName(jdbcDriverClass);
+            if (connection != null) {
+                connection.close();
```
Comment (Shtramak) [inline src/main/java/juja/sqlcmd/DatabaseManager.java]: Maybe it's better to notify the client somehow that already there is unclosed connection? That's the moment when you provide additional logic to your implementation without requirements in a task. (http://bit.ly/2jRFHEn)

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 5]
```diff
@@ -5,16 +5,33 @@
...
+            }
+
             connection = DriverManager.getConnection(
-                    "jdbc:postgresql://localhost:5432" + database, user, password);
+                    jdbcUrl + database, user, password);
             return true;
         } catch (ClassNotFoundException | SQLException e) {
             e.printStackTrace();
+            connection = null;
```
Comment (Shtramak) [inline src/main/java/juja/sqlcmd/DatabaseManager.java]: What if connection is not closed at this moment?

File: src/test/java/juja/sqlcmd/DatabaseManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    private PrintStream originalOut;
+    private PrintStream originalErr;
+
+    @Before
+    public void setUpStreamsAndDbManager() {
+        originalOut = System.out;
+        originalErr = System.err;
+        System.setOut(new PrintStream(outContent));
+        System.setErr(new PrintStream(errContent));
+        databaseManager = new DatabaseManager();
```
Comment (Shtramak) [inline src/test/java/juja/sqlcmd/DatabaseManagerTest.java:26]: databaseManager = new DatabaseManager(); is the only line you need in this method. Why you change streams from System? And then it's better to rename this method to init() or setup()

File: src/test/java/juja/sqlcmd/DatabaseManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @After
+    public void cleanUpStreams() {
```
Comment (Shtramak) [inline src/test/java/juja/sqlcmd/DatabaseManagerTest.java:30]: You don't need this method after deleting unnesessary lines from setUpStreamsAndDbManager()

**Review by Shtramak — APPROVED** [review]

File: src/main/java/juja/sqlcmd/DatabaseManager.java (hunk 2 above)
Comment (kortov) [reply src/main/java/juja/sqlcmd/DatabaseManager.java]: It's just for testing of bad url and driverName)

File: src/test/java/juja/sqlcmd/DatabaseManagerTest.java (hunk 6 above)
Comment (kortov) [reply src/test/java/juja/sqlcmd/DatabaseManagerTest.java:26]: There will be many stacktraces in tests of DbManager constructor, I just hide them

Comment (kortov) [reply src/test/java/juja/sqlcmd/DatabaseManagerTest.java:26]: About init method :) https://youtu.be/otrfSgeK3JI?t=601

File: src/main/java/juja/sqlcmd/DatabaseManager.java (hunk 5 above)
Comment (kortov) [reply src/main/java/juja/sqlcmd/DatabaseManager.java]: Good question, I just copied it from Baglai's example)

File: src/main/java/juja/sqlcmd/DatabaseManager.java (hunk 4 above)
Comment (kortov) [reply src/main/java/juja/sqlcmd/DatabaseManager.java]: I just copied it from Baglai's example but actually despite these lines look reasonable, I don't see any hanging connections in db without them. May be it's redundant

**Review by adaleorion — APPROVED** [review]

**Pull request merged**